common/membership/ringpop/service_resolver.go

```diff
@@ -314,14 +313,57 @@ func (r *serviceResolver) refreshLocked() (*membership.ChangedEvent, error) {
 	return changedEvent, nil
 }
 
-func (r *serviceResolver) getReachableMembers() ([]*hostInfo, error) {
+func (r *serviceResolver) scheduleRefresh(nextEvent int64) {
+	if nextEvent == 0 {
+		return
+	}
+	if _, ok := r.scheduledRefreshMap[nextEvent]; ok {
+		return // already have a timer scheduled for this time
+	}
+	nextEventTime := time.Unix(nextEvent, 0)
+	r.scheduledRefreshMap[nextEvent] = time.AfterFunc(time.Until(nextEventTime), func() {
+		// force refresh asap
+		r.refresh(true)
+		// clean up map
+		r.refreshLock.Lock()
+		defer r.refreshLock.Unlock()
+		delete(r.scheduledRefreshMap, nextEvent)
+	})
+	r.logger.Debug("Membership will refresh at scheduled event", tag.Timestamp(nextEventTime))
+}
+
+func (r *serviceResolver) getReachableMembers() ([]*hostInfo, int64, error) {
 	members, err := r.rp.GetReachableMemberObjects(swim.MemberWithLabelAndValue(roleKey, string(r.service)))
 	if err != nil {
-		return nil, err
+		return nil, 0, err
 	}
 
-	hosts := make([]*hostInfo, len(members))
-	for i, member := range members {
+	nowUnix := time.Now().Unix()
+	// Keep track of earliest scheduled event. We only need to keep track of one, when we
+	// refresh after that one we'll find the next one.
+	nextEvent := int64(math.MaxInt64)
+
+	hosts := make([]*hostInfo, 0, len(members))
+	for _, member := range members {
```

Comment:
This loop does plenty of things now, which makes it difficult for me to understand quickly. Everything is merged to iterate over members only once. If we iterate twice, or maybe even filter and iterate, it will become apparent which `nextEvent` we are interested in and how we build `hosts` slice. Do we need to optimize the number of iterations?

Reply:
It's not quite as simple as filter and then extract and then map, because we may need information from some of the elements we dropped in the filter.

I think we can do filter+extract in one loop and map in another loop though. Is that better? I don't think optimizing the number of iterations makes a difference, but I would like to avoid parsing all the numbers twice.

We always need both values so I don't think splitting the function helps.

Reply:
I share some of the concerns of Sasha about the readability of this loop. What if we refactored it so that the callsite went from this:

```go
hosts, nextEvent, err := r.getReachableMembers()
```

to something like this:

```go
members, err := r.getAllMembers()
if err != nil { 
  return nil, err 
}

nextScheduledMembershipChangeTime := getNextScheduledMembershipChangeTime(members)
r.maybeScheduleRefreshFor(nextScheduledMembershipChangeTime)

r.removeInactiveMembers(&members)
// ... rest of code
```